### PR TITLE
Cr 3175 azure dd scenarios

### PR DESCRIPTION
**Pull request by Tomerva**

### CODE REVIEW

File: cloudrail/knowledge/context/azure/resources/databases/azure_cosmos_db_account.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
+from typing import Optional, List, Dict
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/resources/databases/azure_cosmos_db_account.py:1]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/azure/resources_builders/scanner/cosmos_db_account_builder.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
+from typing import List
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/resources_builders/scanner/cosmos_db_account_builder.py:1]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/azure/resources_builders/terraform/cosmos_db_account_builder.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+from cloudrail.knowledge.context.azure.resources.constants.azure_resource_type import AzureResourceType
```
Comment (Tomerva) [inline cloudrail/knowledge/context/azure/resources_builders/terraform/cosmos_db_account_builder.py:1]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/base_context_builders/base_terraform_builder.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 from cloudrail.knowledge.context.iac_action_type import IacActionType
```
Comment (Tomerva) [inline cloudrail/knowledge/context/base_context_builders/base_terraform_builder.py:9]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/environment_context/business_logic/dependency_invocation.py [hunk 5]
```diff
@@ -6,6 +6,7 @@
 from cloudrail.knowledge.exceptions import ContextEnrichmentException, ResourceDependencyNotFoundException
```
Comment (Tomerva) [inline cloudrail/knowledge/context/environment_context/business_logic/dependency_invocation.py:6]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/parallel/process/async_queue_executor.py [hunk 6]
```diff
@@ -4,6 +4,8 @@
 from multiprocessing import Process, connection
```
Comment (Tomerva) [inline cloudrail/knowledge/context/parallel/process/async_queue_executor.py:4]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/utils/log_utils.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
+import json
```
Comment (Tomerva) [inline cloudrail/knowledge/utils/log_utils.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_indirect_public_access_db_rds_rule.py [hunk 8]
```diff
@@ -1,5 +1,6 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_indirect_public_access_db_rds_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_indirect_public_access_db_redshift_rule.py [hunk 9]
```diff
@@ -1,5 +1,6 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_indirect_public_access_db_redshift_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_neptune_rule.py [hunk 10]
```diff
@@ -1,6 +1,7 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_neptune_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_rds_rule.py [hunk 11]
```diff
@@ -1,6 +1,7 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_rds_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_redshift_rule.py [hunk 12]
```diff
@@ -1,6 +1,7 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_db_redshift_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_dms_replication_instance_rule.py [hunk 13]
```diff
@@ -1,6 +1,7 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_dms_replication_instance_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_eks_api_rule.py [hunk 14]
```diff
@@ -1,6 +1,7 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_eks_api_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_security_groups_port_rule.py [hunk 15]
```diff
@@ -75,4 +75,5 @@ def _create_violating_context():
         all_inclusive_sg_rule.property_value = '0.0.0.0/0'
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_public_access_security_groups_port_rule.py:75]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_s3_acl_allow_public_access_rule.py [hunk 16]
```diff
@@ -19,7 +19,7 @@ def test_s3_acl_disallow_public_and_cross_account_fail(self):
         bucket = create_empty_entity(S3Bucket, bucket_name='bucket_name')
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/public_access_validation_rules/test_s3_acl_allow_public_access_rule.py:19]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/test_ec2_role_share_rule.py [hunk 17]
```diff
@@ -32,7 +32,8 @@ def test_ec2_role_share_rule_fail(self):
         public_ec2.iam_profile_name = 'iam_profile_name'
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/test_ec2_role_share_rule.py:32]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/test_ensure_all_used_default_security_groups_restrict_all_traffic_rule.py [hunk 18]
```diff
@@ -1,5 +1,6 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/test_ensure_all_used_default_security_groups_restrict_all_traffic_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/context_aware/test_ensure_iam_entities_policy_managed_solely_rule.py [hunk 19]
```diff
@@ -1,5 +1,6 @@
 import unittest
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/context_aware/test_ensure_iam_entities_policy_managed_solely_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: tests/knowledge/rules/aws/non_context_aware/log_validation_rules/test_ensure_lambda_function_has_non_infinite_log_retention_rule.py [hunk 20]
```diff
@@ -19,7 +19,7 @@ def test_non_car_lambda_logging_not_infinite_fail(self):
         log_group: CloudWatchLogGroup = create_empty_entity(CloudWatchLogGroup)
```
Comment (Tomerva) [inline tests/knowledge/rules/aws/non_context_aware/log_validation_rules/test_ensure_lambda_function_has_non_infinite_log_retention_rule.py:19]: dont know why its showing as a change - this is from merging from develop

File: tests/unit/test_base_rule.py [hunk 21]
```diff
@@ -0,0 +1,65 @@
+import unittest
```
Comment (Tomerva) [inline tests/unit/test_base_rule.py:1]: dont know why its showing as a change - this is from merging from develop

File: cloudrail/knowledge/context/environment_context/business_logic/resource_invalidator.py [hunk 22]
```diff
@@ -124,25 +123,39 @@ def _mergeable_attributes(mergeable):
...
+    @staticmethod
+    def _get_by_iterable_logic(resources: Union[AliasesDict[_TMergeable], Iterable, Dict[str, _TMergeable]],
+                               resource_id: str,
+                               case_sensitive: bool) -> Optional[_TMergeable]:
+        try:
+            if case_sensitive:
+                return next((resource for resource in resources
+                             if resource_id == resource.get_id() or resource_id in resource.aliases), None)
+            else:
+                return next((resource for resource in resources
```
Comment (Imanuel-Mizrahi) [inline cloudrail/knowledge/context/environment_context/business_logic/resource_invalidator.py:153]: Why not always compare with lower()?

**Review by Imanuel-Mizrahi — APPROVED** [review]

File: cloudrail/knowledge/context/environment_context/business_logic/resource_invalidator.py (hunk 22 above)
Comment (Tomerva) [reply cloudrail/knowledge/context/environment_context/business_logic/resource_invalidator.py:153]: performane tweak

**Review by Viperain — APPROVED** [review]

**Pull request merged**